lib/widgets/nft_collection_bloc.dart

```diff
@@ -113,63 +200,138 @@ class NftCollectionBloc
-            NftCollection.logger
-                .info("[Stream.refreshTokensInIsolate] getEvent Done");
-            add(_SubRefreshTokensEvent(NftLoadingState.done));
-          });
-        }
+          if (event.isNotEmpty) {
+            NftCollection.logger.info(
+                "[Stream.refreshTokensInIsolate] UpdateTokensEvent ${event.length} tokens");
+            if (state.nextKey?.offset != null) {
+              final addingTokens = event
+                  .where(
+                    (element) =>
+                        element.lastActivityTime.millisecondsSinceEpoch >=
+                        state.nextKey!.offset!,
+                  )
+                  .toList();
+              add(UpdateTokensEvent(
+                  state: NftLoadingState.loading, tokens: addingTokens));
+            }
+          }
+        }, onDone: () async {
+          NftCollection.logger
+              .info("[Stream.refreshTokensInIsolate] getEvent Done");
+          add(UpdateTokensEvent(state: NftLoadingState.done));
+        });
       } catch (exception) {
+        add(UpdateTokensEvent(state: NftLoadingState.error));
+
         NftCollection.logger.warning("Error: $exception");
       }
     });
 
+    on<RefreshNftCollectionByIDs>((event, emit) async {
+      NftCollection.logger
+          .info("[NftCollectionBloc] RefreshNftCollectionByIDs");
+      if (event.debugTokenIds?.isNotEmpty ?? false) {
+        _debugTokenIds = event.debugTokenIds ?? [];
+        fetchManuallyTokens(_debugTokenIds);
+      }
+      if (event.ids?.isEmpty ?? true) {
+        emit(state.copyWith(
+          nextKey: state.nextKey,
+          tokens: [],
+          state: NftLoadingState.done,
+        ));
+        return;
+      }
+
+      final assetTokens =
+          await database.assetTokenDao.findAllAssetTokensByTokenIDs(event.ids!);
+      emit(state.copyWith(
+        nextKey: state.nextKey,
+        tokens: assetTokens,
+        state: NftLoadingState.done,
+      ));
+    });
+
+    on<UpdateTokensEvent>((event, emit) async {
+      final tokens = [...event.tokens, ...state.tokens]
+          .unique((element) => element.id + element.owner);
+
+      tokens?.removeWhere((element) =>
+          !activeAddress.contains(element.owner) && element.isDebugged != true);
+
+      emit(
+        state.copyWith(
+          state: event.state,
+          tokens: tokens,
+          nextKey: state.nextKey,
+        ),
+      );
+    });
+
+    on<ReloadEvent>((event, emit) async {
+      emit(state.copyWith(nextKey: state.nextKey));
+    });
+
     on<RequestIndexEvent>((event, emit) async {
       tokensService.reindexAddresses(_filterAddresses(event.addresses));
     });
+  }
+  Map<int, List<String>> mapAddressesByLastRefreshedTime(
+      List<AddressIndex> addresses, DateTime? lastRefreshedTime) {
+    if (addresses.isEmpty) return {};
+    final listAddresses = addresses.map((e) => e.address).toList();
+    if (lastRefreshedTime == null) {
```

Comment:
do you miss the the case to add the `listPendingAddresses` into the address list when `lastRefreshedTime == null`